core/src/main/java/org/apache/iceberg/rest/HttpRESTClient.java

```diff
@@ -0,0 +1,288 @@
+    } catch (IOException | ParseException e) {
+      if (responseType != null) {
+        LOG.warn("Failed to convert HTTP response body to string", e);
+      }
+      return Optional.empty();
+    }
+  }
+
+  // Per the spec, the only currently defined / used "success" responses are 200 and 202.
+  private static boolean isSuccessful(CloseableHttpResponse response) {
+    int code = response.getCode();
+    return code ==  HttpStatus.SC_OK || code == HttpStatus.SC_ACCEPTED;
+  }
+
+  private static ErrorResponse buildDefaultErrorResponse(CloseableHttpResponse response) {
+    String responseReason = response.getReasonPhrase();
+    String message =
+        responseReason != null && !responseReason.isEmpty() ? responseReason :
+            EnglishReasonPhraseCatalog.INSTANCE.getReason(response.getCode(), null /* ignored */);
+    String type = "ClientGeneratedErrorResponse - No error response body could be parsed";
+    return ErrorResponse.builder()
+        .responseCode(response.getCode())
+        .withMessage(message)
+        .withType(type)
+        .build();
+  }
+
+  private static ErrorResponse parseOrCreateErrorResponse(
+      CloseableHttpResponse response, Optional<String> responseBody, Method method, Class<?> responseType) {
+    try {
+      return responseBody.map(ErrorResponseParser::fromJson).orElseGet(() -> buildDefaultErrorResponse(response));
+    } catch (Exception e) {
+      // Only log a warning that we couldn't parse an error response if some response body was expected
+      if (responseType != null) {
+        LOG.warn(
+            "Failed to parse standard error response for {} call expecting {} response. Creating one instead",
+            method, responseType.getSimpleName(), e);
+      }
+
+      return buildDefaultErrorResponse(response);
+    }
+  }
+
+  /**
+   * Method to execute an HTTP request and process the corresponding response.
+   *
+   * @param method       - HTTP method, such as GET, POST, HEAD, etc.
+   * @param path         - URL path to send the request to
+   * @param requestBody  - Content to place in the request body
+   * @param responseType - Class of the Response type. Needs to have serializer registered with ObjectMapper
+   * @param errorHandler - Error handler delegated for HTTP responses which handles server error responses
+   * @param <T>          - Class type of the response for deserialization. Must be registered with the ObjectMapper.
+   * @return The response entity, parsed and converted to its type T
+   */
+  @Nullable
+  public <T> T execute(
+      Method method, String path, Object requestBody, Class<T> responseType, Consumer<ErrorResponse> errorHandler) {
+    if (path.startsWith("/")) {
+      throw new RESTException(
+          "Received a malformed path for a REST request: %s. Paths should not start with /", path);
+    }
+
+    // TODO - Add RESTUtil.stripTrailingSlash call on uri once PR #4244 is merged and update associated RESTClient
```

Comment:
This was merged.

Also, do we need to call `stripTrailingSlash`? Since this is only going to be used with paths constructed by this library, I think we can simply fail if the path ends in `/`.

Reply:
Wait, this does throw an exception above. What is the problem that this needs `stripTrailingSlash` for?

Reply:
The reason I had been thinking to handle the slashes was that in the `mockserver` based tests, to build the request, the path needs to start with `/`.

But in our code, we're passing in the path without `/`. I was going to strip trialing slashes from the URI and then require the passed in path _does_ have a `/`. But I'm not sure I'm crazy about that change either way.

Let me toy with it for a moment and see if something better can be done to make it simpler and less error prone.

Reply:
Actually, I was mistaken. I needed `stripTrailingSlash` for the `uri` instead.

I've added that into the builder so that the uri, if passed as `localhost:1080/` or similar works fine.